### ISSUE TITLE
Start Pulsar WSGI Server from Docker

### ISSUE THREAD

**Issue by MrLoh** [quantmind/pulsar#241]
Hey, I am using Pulsar for JSON-RPC and everything works fine locally, but I can't send calls from outside the Docker container, where the WSGI Server is running. Do I need to expose other ports then the one that I expose with `--bind` or what could be the issue?

I was hoping I could simply expose a JSON-RPC API with pulsar, like I do it with flask and consume that with other pulsar or node apps.

**Comment by lsbardel** [comment]
That should just work.
What do you bind to? Try the bind to all IP addresses (``--bind :8060`` for example)
Maybe the problem is in your docker configuration?

**Comment by IwanLD** [comment]
# TLDR

Hey, I'm having the same issue, but found a hacky workaround - see SOLUTION beneath if your tldr ;) 

# ANALYSIS

To dive into it I have the simple test server

```
def hello(environ, start_response):
    data = b"Hello World!"
    response_headers = [('Content-type','text/plain'),
                        ('Content-Length', str(len(data)))]
    start_response("200 OK", response_headers)
    return [data]

if __name__ == '__main__':
    wsgi.WSGIServer(hello).start()
```

when it starts via ```python3 app.py``` , I get

```
16:40:21 [p=383, t=139677125908224, INFO, pulsar.arbiter] mailbox serving on 127.0.0.1:42675
16:40:21 [p=383, t=139677125908224, INFO, pulsar.arbiter] started
16:40:21 [p=383, t=139677125908224, INFO, pulsar.wsgi] started
16:40:21 [p=385, t=139677125908224, INFO, pulsar.wsgi.worker] wsgi serving on 127.0.0.1:8060
16:40:21 [p=385, t=139677125908224, INFO, pulsar.wsgi.worker] started
```
My docker is basically run with ```docker -p 8060:8060 -p 5000:5000 -it myimage```.  So the port 8060 is open. (5000 is just to show that flask works later)

Inside the docker
```
$ curl localhost:8060
Hello World!
```

Outside the docker
```
$ curl localhost:8060
curl: (52) Empty reply from server
$ curl localhost:5000
curl: (52) Empty reply from server
$ curl localhost:5001
curl: (7) Failed to connect to localhost port 5001: Connection refused
```
When I then run a hello world in flask on 5000 everithing works fine.
Outside the docker
```
$ curl localhost:5000
Hello world
```

So it seems that somehow to the outsite world it is if there were no one listening to the 8060 port.

Now when I run it with `python3 app.py --bind localhost:8060` the output is a bit different...

```
17:27:46 [p=1056, t=140084304557824, INFO, pulsar.arbiter] mailbox serving on 127.0.0.1:43803
17:27:46 [p=1056, t=140084304557824, INFO, pulsar.arbiter] started
17:27:46 [p=1056, t=140084304557824, INFO, pulsar.wsgi] started
17:27:46 [p=1059, t=140084304557824, INFO, pulsar.wsgi.worker] wsgi serving on [::1]:8060
17:27:46 [p=1059, t=140084304557824, INFO, pulsar.wsgi.worker] wsgi serving on 127.0.0.1:8060
17:27:46 [p=1059, t=140084304557824, INFO, pulsar.wsgi.worker] started
```

... but the result stays the same.

# SOLUTION

Now when inside the docker I run 

```
$ ip route show
default via 172.17.0.1 dev eth0
172.17.0.0/16 dev eth0  proto kernel  scope link  src 172.17.0.2
```

Now when I run it with `python3 app.py --bind 172.17.0.2:8060` there comes remedy.

On the outside I get
```
$ curl localhost:8060
Hello World!
```

Yeay!

On the inside of the container I get

```
$ curl localhost:8060
curl: (7) Failed to connect to localhost port 8060: Connection refused
```

and even

```
$ curl 172.17.0.2:8060
curl: (7) Failed to connect to 172.17.0.2 port 8060: Connection refused
```

but luckily

```
$ curl 172.17.0.1:8060
Hello World!
```

# ANALYSIS

I'm no expert on networks but so far my interpretation is this: Docker seems to set up some sort of extra routes layer between the containers localhost and the outer world.

Every other web framework I know (flask, node, webapp2, apache, ...) binds both localhost and this extra layer by default and works with docker out of the box.

Pulsar so far binds only either or. Imho, this should be either changed (to always binding both) or added as an extra option.

# FURTHER READING

I got the idea from this [SO post]
(http://stackoverflow.com/questions/31324981/how-to-access-host-port-from-docker-container)

**Comment by MrLoh** [comment]
It seems like simply doing the `--bind :8060` also does the trick. I can then connect inside the docker to `localhost:8060` and outside the docker to `docker_ip:mapped_port`, just as expected.

**Comment by IwanLD** [comment]
I'm a bit confused, that this got closed without a comment. I consider this a major gotcha, since docker is gaining popularity quickly. Especially since a lot of the pulsar documentaion (well the docstrings actually) contains examples with "bind 127.0.0.1". It should be at least noted somewhere (preferably somewhere besides this post).

**Comment by lsbardel** [comment]
I've added some info in the docs, commit above.
Available on the dev documentation http://quantmind.github.io/pulsar/apps/socket.html
